pyccel/ast/headers.py

```diff
@@ -204,15 +252,41 @@ def build_argument(var_name, dc):
                         is_argument=True)
             return var
 
+        def process_template(signature, Tname, d_type):
+            #Replaces templates named Tname inside signature, with the given type.
+            new_sig = tuple(d_type if 'datatype' in t and t['datatype'] == Tname\
+                    else t for t in signature)
+            return new_sig
+
+        def find_templates(signature, templates):
+            #Creates a dictionary of only used templates in signature.
+            new_templates = {d_type['datatype']:templates[d_type['datatype']]\
+                    for d_type in signature\
+                    if 'datatype' in d_type and d_type['datatype'] in templates}
+            return new_templates
 
         for i in self.dtypes:
             if isinstance(i, UnionType):
+                for d_type in i.args:
+                    if d_type['datatype'] in templates:
+                        errors.report(TEMPLATE_IN_UNIONTYPE,
```

Comment:
Do we have an object that we could provide as a symbol here to help the user identify which line is causing the error?

Reply:
can we use the name of the function which header causes the problem ?

Reply:
That seems like it would be sufficient if you don't have the header object to print

Reply:
I have it, but the problem the header contains dictionaries of types, which makes the error hard to read

Reply:
Up to you then. It might be worth checking if the dictionary is printed when you print the header. If you think it's hard to read. The function should be sufficient

Reply:
I checked, the dictionaries are printed, I will go with the function's name